librad/src/uri.rs

```diff
@@ -519,14 +521,41 @@ impl<'de> Deserialize<'de> for RadUrl {
     }
 }
 
+pub struct RadUrlRef<'a> {
+    pub authority: &'a PeerId,
+    pub urn: &'a RadUrn,
+}
+
+impl<'a> RadUrlRef<'a> {
+    pub fn to_owned(&self) -> RadUrl {
```

Comment:
You could probably add a https://doc.rust-lang.org/std/borrow/trait.ToOwned.html impl for this.

Reply:
I tried that, but couldn't get it to typecheck. I might be missing something, but the issue seems to be: `ToOwned` has a constraint that the associated type `Owned` must be `Borrow`. Now, `Borrow` is parametrised over the borrowed type, and `fn borrow(&self) -> &Borrowed`. There is no way I can construct a `Borrowed` and then lend it out -- `borrow` owns it.

Since I see this `*Ref` pattern with inherent methods being used all over the place in Rust, I figure this _is_ actually an issue. Or I'm not interrigent enough, but either way I lost patience after a while :grimacing:

Reply:
Fair enough, I think I'd have to give it a go myself to understand better. But we shouldn't waste both of our time :smile: